pygenometracks/utilities.py

```diff
@@ -194,35 +194,36 @@ def plot_coverage(ax, x_values, score_list, plot_type, size, color,
                             facecolor=negative_color,
                             alpha=alpha)
 
-    def transform(score_list, transform, file):
-        if transform == 'no':
-            return(score_list)
-        elif transform == 'log':
-            if score_list.min() < 0:
-                msg = ("\n*ERROR*\ncoverage contains negative values.\n"
-                       "log(<values>) transformation can not be applied to \n"
-                       "values in file: {}".format(file))
-                raise Exception(msg)
-            else:
-                return(np.log(score_list))
-        elif transform == 'log1p':
-            if score_list.min() < 1:
-                msg = ("\n*ERROR*\ncoverage contains values below 1.\n"
-                       "log1p(<values>) transformation can not be applied to \n"
-                       "values in file: {}".format(file))
-                raise Exception(msg)
-            else:
-                return(np.log(score_list + 1))
-        elif transform == '-log':
-            if score_list.max() > 0:
-                msg = ("\n*ERROR*\ncoverage contains positive values.\n"
-                       "-log(<values>) transformation can not be applied to \n"
-                       "values in file: {}".format(file))
-                raise Exception(msg)
-            else:
-                return(np.log(- score_list))
+
+def transform(score_list, transform, file):
+    if transform == 'no':
+        return(score_list)
+    elif transform == 'log':
+        if np.nanmin(score_list) <= 0:
+            msg = ("\n*ERROR*\ncoverage contains negative or zero values.\n"
+                    "log(<values>) transformation can not be applied to \n"
+                    "values in file: {}".format(file))
+            raise Exception(msg)
+        else:
+            return(np.log(score_list))
+    elif transform == 'log1p':
+        if np.nanmin(score_list) <= - 1:
+            msg = ("\n*ERROR*\ncoverage contains values below or equal to 1.\n"
+                    "log1p(<values>) transformation can not be applied to \n"
+                    "values in file: {}".format(file))
+            raise Exception(msg)
         else:
-            warnings.warn('The transform: {} for file {} is not valid.'
-                          'will not use any transformation'.format(transform,
-                                                                   file))
-            return(score_list)
+            return(np.log(score_list + 1))
+    elif transform == '-log':
+        if np.nanmax(score_list.max) >= 0:
+            msg = ("\n*ERROR*\ncoverage contains positive or zero values.\n"
```

Comment:
dont you think that coverage may have 0?

Reply:
Hi,
I don't know what should be done:
- option 1: We do not apply transformation when data are not in the correct range
- option 2: We put to nan the values that would give infinite values.
I used option 1 because this was what was done for HiC:
https://github.com/deeptools/pyGenomeTracks/blob/d25d321e0e9b519224205ac16c0e410806afe4f1/pygenometracks/tracks/HiCMatrixTrack.py#L117-L132
But we can decide to change both to option 2. No problem for me.

Reply:
you can allow for log2 or/and log10 with a user-defined pseudo count.

Reply:
Yes, that's true. I was just thinking about doing the same as in HiC that's why I just put log/-log/log1p...

Reply:
I think there is no point to use log2 or log10 as they are all proportional and the scale will still give the original value but I could add a log_pseudocount which is used with log and -log.
I think we still need to choose an option (raise an error or put nan).

Reply:
we definitely need a pseudo count. 
How do you mean by there is `no use` and `still give the original value`? log2 means the values are transformed by numpy.log2 and log10 means they get transformed by numpy.log10, this can be decided by user.  How can these values be the same as the original ones?
why would it be any nan? If there is any, it can throw an error for now, but i cannot see that in the context of gene expression if you don't have 0.

Reply:
> we definitely need a pseudo count.

There was log1p with a pseudo count of 1 but I will add log_pseudocount, no issue with that.

- Why did I say there was no difference between log2 and log10?
Here is an example of output with no transformation or with log1p:
![master_log1p](https://user-images.githubusercontent.com/30404086/70801238-a819ce80-1dae-11ea-9eef-1469fd0c81bb.png)
The values on the y axis are not the log1p values but the original values. So if I replot the same example with log1p or log2(1+values) or log10(1+values) I will always have the same result because log2(1+values) = log(1+values)/log(2) and the same way log10(1+values) = log(1 + values)/log(10). I think the idea of the original issue was to have a a y-axis with log scale but maybe I misunderstood.

- It is possible to have tracks with only NA and no 0 values. For example, I have a 4Cseq track which is like this:
![master_log](https://user-images.githubusercontent.com/30404086/70801655-ba483c80-1daf-11ea-97e3-33a8043f34ad.png)
But I agree this is quite rare but you could have a track with ratio for which you would like to do a log without pseudocount.

Reply:
>     * I think the idea of the original issue was to have a a y-axis with log scale but maybe I misunderstood.

yes that is what they have asked for.

Reply:
>  a pseudo count of 1 but I will add log_pseudocount

+1 was an example, that is why i have suggested to make it a user define parameter with some default value.

Reply:
I think it is a good idea to have log10 and log2 on top the natural log that you have already implemented or even instead of it.

Reply:
If I understand well, you would prefer to put in y-axis the transformed value, not the original value, right?
I just noticed that it was depending of browser.
For example in igv keeps original values:
![image](https://user-images.githubusercontent.com/30404086/70806462-aa365a00-1dbb-11ea-8777-c52556320dd0.png)
But UCSC put tranformed:
![image](https://user-images.githubusercontent.com/30404086/70806753-4ceed880-1dbc-11ea-8d06-cb62c2ce0e45.png)
Let's put another parameter: y_axis_values = 'original' or 'transformed'.
And I add 'log2' 'log10' as possible values for transform
Would it be ok for you?

Reply:
sure

Reply:
Here is the new output:
![master_log](https://user-images.githubusercontent.com/30404086/70903689-b01c7d00-1fff-11ea-9200-dbf10ed48c23.png)